rfc/custom-modules-v1.md

```diff
@@ -0,0 +1,706 @@
+	// The client name of the operation
+	ClientName string
+	// The client version of the operation
+	ClientVersion string
+	// The uploaded files of the operation
+	UploadedFiles map[string]*core.UploadedFile
+	// The plan of the operation
+	Plan *core.Plan
+	// The normalization of the operation
+	Normalization *core.Normalization
+}
+
+type RouterRequest struct {
+	// The original GraphQL request with all the information like query, variables, operation name, extensions etc.
+	Request *core.GraphQLRequest
+	// The parsed, normaliazed and planned operation with all the information like name, variables, type, document representation,
+	// client name version, uploaded files, plan, normalization, persisted operation etc.
+	Operation *core.Operation
+	// The active Telemetry instance
+	Telemetry *core.Telemetry
+	// The active graph
+    Federatedgraph *core.Federatedgraph
+	// The original HTTP request
+	HttpRequest *http.Request
+	// Logger for the router request
+	Logger *zap.Logger
+}
+
+type RouterResponse struct {
+	// The original RouterRequest
+	Request *core.RouterRequest
+	// The active Telemetry instance
+	Telemetry *core.Telemetry
+	// The final GraphQL response with all the information like data, errors, extensions etc.
+	// This is the response that will be sent to the client and can be manipulated or replaced
+	Response *core.GraphQLResponse
+	// The original HTTP response
+	HttpResponse *http.Response
+}
+
+type SubgraphRequest struct {
+	// The active subgraph
+	Subgraph *core.Subgraph
+	// The original Router request
+	RouterRequest *core.RouterRequest
+	// The final GraphQL request to the subgraph
+	Request *core.GraphQLRequest
+	// The active Telemetry instance
+	Telemetry *core.Telemetry
+	// The original http request to the subgraph
+	HttpRequest *http.Request
+	// Logger for the subgraph request
+	Logger *zap.Logger
+}
+
+type SubgraphResponse struct {
+	// The active subgraph
+	Subgraph *core.Subgraph
+	// The original Router request
+	RouterRequest *core.RouterRequest
+	// The active Telemetry instance
+	Telemetry *core.Telemetry
+	// The final GraphQL response from the subgraph
+	Response *core.GraphQLResponse
+	// The original http response from the subgraph
+	HttpResponse *http.Response
+}
+
+// Router Hooks
+
+type RouterRequestHook interface {
+	// OnRouterRequest is called when a request is made to the router and after all GraphQL information is available
+	// Returning an error will result in a GraphQL error being returned to the client.
+	OnRouterRequest(ctx *core.RouterRequest, err error) error
+}
+
+type RouterResponseHook interface {
+	// OnRouterResponse is called before the response is sent to the client
+	// Returning an error will result in a GraphQL error being returned to the client.
+	OnRouterResponse(ctx *core.RouterResponse, err error) error
+}
+
+type RouterErrorHook interface {
+	// OnError is called when an error occurs during the router lifecycle
+	OnRouterError(err error)
```

Comment:
I'm not sure how to use it yet. I would like to be able to replace the status code in the subgraph responses, to decide which are good and which are not, which should be retrained and which should not. And then modify the router's final response after gluing and downloading everything

Reply:
This interface will have to be driven to your error type somehow? Or how to change it?

Reply:
It seems that what I described will be done in OnRouterResponse https://github.com/wundergraph/cosmo/pull/1063/files#r1778364936? And this method will be triggered after OnRouterResponse?